includes/REST/ProductControllerV3.php

```
@@ -0,0 +1,165 @@
+
+    /**
+     * Resolve product request body before schema validation runs (so WC schema sees WC-shaped payload).
+     * Runs on rest_pre_dispatch so that schema_to_wc_api is applied before args validation.
+     *
+     * @param mixed            $result  Response to replace the short-circuit result with.
+     * @param WP_REST_Server  $server  Server instance.
+     * @param WP_REST_Request $request Request used to generate the response.
+     *
+     * @return mixed Unchanged result so dispatch continues; request body is modified in place.
+     */
+    public function resolve_product_payload_before_validation( $result, $server, $request ) {
```

PHPMD (2.15.0)

[warning] Avoid unused parameters such as '$server'. (undefined)

(UnusedFormalParameter)